File: tools/sentinel/src/lib/tools/butler.ts
import fs from 'fs/promises';
import path from 'path';
import { execa } from 'execa';

// Get the root of the entire vault (two levels up from tools/sentinel)
const getProjectRoot = (): string => {
    // __dirname logic equivalent for ES modules running via tsx
    return path.resolve(process.cwd(), '../..');
};

/**
 * 自动归档历史日记到 YYYY/MM 目录
 */
async function archiveDiary(projectRoot: string): Promise<number> {
    const historyDir = path.join(projectRoot, 'history');
    let movedCount = 0;

    try {
        const stats = await fs.stat(historyDir);
        if (!stats.isDirectory()) return 0;
    } catch {
        return 0; // Dir doesn't exist
    }

    const todayStr = new Date().toISOString().split('T')[0];
    const todayFile = `${todayStr}.md`;
    const excludeFiles = ['日记模版.md', todayFile];

    const files = await fs.readdir(historyDir);

    for (const filename of files) {
        if (!filename.endsWith('.md') || excludeFiles.includes(filename)) {
            continue;
        }

        // Format: YYYY-MM-DD.md
        if (filename.length === 13 && /^\d{4}-\d{2}/.test(filename)) {
            const year = filename.substring(0, 4);
            const month = filename.substring(5, 7);

            const targetDir = path.join(historyDir, year, month);
            await fs.mkdir(targetDir, { recursive: true });

            const sourcePath = path.join(historyDir, filename);
            const targetPath = path.join(targetDir, filename);

            await fs.rename(sourcePath, targetPath);
            movedCount++;
        }
    }

    return movedCount;
}

/**
 * 清理 inbox 中的空文件和无价值垃圾碎片
 */
async function cleanInbox(projectRoot: string): Promise<number> {
    const inboxDir = path.join(projectRoot, 'inbox');
    let deletedCount = 0;

    try {
        const stats = await fs.stat(inboxDir);
        if (!stats.isDirectory()) return 0;
    } catch {
        return 0; // Dir doesn't exist
    }

    // A simple recursive dir walk for Node.js
    async function walk(dir: string) {
        const entries = await fs.readdir(dir, { withFileTypes: true });
        for (const entry of entries) {
            const fullPath = path.join(dir, entry.name);
            if (entry.isDirectory()) {
                await walk(fullPath);
            } else {
                // Skip system files
                if (entry.name === '.DS_Store' || entry.name.startsWith('._')) {
                    continue;
                }

                try {
                    const stats = await fs.stat(fullPath);
                    if (stats.size === 0) {
                        await fs.unlink(fullPath);
                        deletedCount++;
                    } else if (entry.name.endsWith('.md') && stats.size < 10) {
                        const content = await fs.readFile(fullPath, 'utf-8');
                        if (!content.trim()) {
                            await fs.unlink(fullPath);
                            deletedCount++;
                        }
                    }
                } catch (e) {
                    console.error(`[Butler] Error processing file ${entry.name}:`, e);
                }
            }
        }
    }

    await walk(inboxDir);
    return deletedCount;
}

/**
 * 将变更提交至 Git
 */
async function gitCommit(projectRoot: string): Promise<string> {
    try {
        // Run git status to see if anything changed
        const { stdout: statusOut } = await execa('git', ['status', '--porcelain'], { cwd: projectRoot });

        if (!statusOut.trim()) {
            return '✓ 工作区干净，无需提交。';
        }

        await execa('git', ['add', 'history/'], { cwd: projectRoot });
        await execa('git', ['add', 'inbox/'], { cwd: projectRoot });

        // Check if there are cached changes
        try {
            await execa('git', ['diff', '--cached', '--quiet'], { cwd: projectRoot });
            // If exit code is 0, no changes were added
            return '✓ history 和 inbox 目录下无实质变更。';
        } catch (e: any) {
            // Exit code 1 means there are changes!
            if (e.exitCode === 1) {
                await execa('git', ['commit', '-m', 'chore: 🤖 管家日常清扫'], { cwd: projectRoot });
                return '✅ 已成功固化管家清扫的里程碑版本。';
            }
            throw e;
        }

    } catch (e: any) {
        return `❌ Git 执行失败: ${e.message}`;
    }
}

/**
 * 暴露给外部 (Telegram / CLI) 的聚合入口
 */
export async function runMaintenance(): Promise<string> {
    try {
        const projectRoot = getProjectRoot();
        const logs: string[] = [];

        const timeStr = new Date().toLocaleString('zh-CN');
        logs.push(`🕰️ [${timeStr}] 🤖 管家开始日常巡检。`);

        const archivedCount = await archiveDiary(projectRoot);
        if (archivedCount > 0) {
            logs.push(`➡️ 归档历史日记: ${archivedCount} 篇`);
        } else {
            logs.push(`✓ 没有需要归档的旧日记。`);
        }

        const deletedCount = await cleanInbox(projectRoot);
        if (deletedCount > 0) {
            logs.push(`❌ 清理 Inbox 垃圾碎片: ${deletedCount} 份`);
        } else {
            logs.push(`✓ Inbox 保持清洁，未发现垃圾。`);
        }

        if (archivedCount > 0 || deletedCount > 0) {
            const commitResult = await gitCommit(projectRoot);
            logs.push(commitResult);
        }

        logs.push(`\n✨ 巡检完毕，您的知识库一尘不染。`);

        return logs.join('\n');
    } catch (error: any) {
        return `❌ 管家巡检过程中发生严重错误: ${error.message || error}`;
    }
}

File: neo-tools/refinery/butler.py
import os
import subprocess
from datetime import datetime
from pathlib import Path
import sys

def get_project_root() -> Path:
    env_dir = os.environ.get("GEMINI_WORK_DIR")
    if env_dir:
        return Path(env_dir)
    return Path(os.getcwd()).parent.parent

def archive_diary(project_root: Path) -> int:
    history_dir = project_root / 'history'
    moved_count = 0
    if not history_dir.is_dir():
        return 0
        
    today_str = datetime.now().strftime("%Y-%m-%d")
    today_file = f"{today_str}.md"
    exclude_files = ['日记模版.md', today_file]
    
    for item in history_dir.iterdir():
        if not item.is_file() or not item.name.endswith('.md') or item.name in exclude_files:
            continue
            
        if len(item.name) == 13 and item.name[:4].isdigit() and item.name[4] == '-' and item.name[5:7].isdigit():
            year = item.name[:4]
            month = item.name[5:7]
            
            target_dir = history_dir / year / month
            target_dir.mkdir(parents=True, exist_ok=True)
            
            target_path = target_dir / item.name
            item.rename(target_path)
            moved_count += 1
            
    return moved_count

def clean_inbox(project_root: Path) -> int:
    inbox_dir = project_root / 'inbox'
    deleted_count = 0
    if not inbox_dir.is_dir():
        return 0
        
    for root, dirs, files in os.walk(inbox_dir):
        for file in files:
            if file == '.DS_Store' or file.startswith('._'):
                continue
                
            full_path = Path(root) / file
            try:
                stats = full_path.stat()
                if stats.st_size == 0:
                    full_path.unlink()
                    deleted_count += 1
                elif file.endswith('.md') and stats.st_size < 10:
                    content = full_path.read_text(encoding='utf-8')
                    if not content.strip():
                        full_path.unlink()
                        deleted_count += 1
            except Exception as e:
                print(f"[Butler] Error processing file {file}: {e}", file=sys.stderr)
                
    return deleted_count

def git_commit(project_root: Path) -> str:
    try:
        status_out = subprocess.run(['git', 'status', '--porcelain'], cwd=project_root, capture_output=True, text=True, check=True).stdout
        if not status_out.strip():
            return '✓ 工作区干净，无需提交。'
            
        subprocess.run(['git', 'add', 'history/'], cwd=project_root, check=True)
        subprocess.run(['git', 'add', 'inbox/'], cwd=project_root, check=True)
        
        diff_res = subprocess.run(['git', 'diff', '--cached', '--quiet'], cwd=project_root)
        if diff_res.returncode == 0:
            return '✓ history 和 inbox 目录下无实质变更。'
        else:
            subprocess.run(['git', 'commit', '-m', 'chore: 🤖 管家日常清扫'], cwd=project_root, check=True)
            return '✅ 已成功固化管家清扫的里程碑版本。'
    except Exception as e:
        return f'❌ Git 执行失败: {e}'

def run_maintenance() -> str:
    try:
        project_root = get_project_root()
        logs = []
        
        time_str = datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        logs.append(f"🕰️ [{time_str}] 🤖 管家开始日常巡检。")
        
        archived_count = archive_diary(project_root)
        if archived_count > 0:
            logs.append(f"➡️ 归档历史日记: {archived_count} 篇")
        else:
            logs.append("✓ 没有需要归档的旧日记。")
            
        deleted_count = clean_inbox(project_root)
        if deleted_count > 0:
            logs.append(f"❌ 清理 Inbox 垃圾碎片: {deleted_count} 份")
        else:
            logs.append("✓ Inbox 保持清洁，未发现垃圾。")
            
        if archived_count > 0 or deleted_count > 0:
            commit_result = git_commit(project_root)
            logs.append(commit_result)
            
        logs.append("\n✨ 巡检完毕，您的知识库一尘不染。")
        return "\n".join(logs)
    except Exception as e:
        return f"❌ 管家巡检过程中发生严重错误: {e}"

if __name__ == '__main__':
    print(run_maintenance())
